### PR TITLE
fix PEP 695 type alias edge cases

### PR DESCRIPTION
**Pull request by youknowone**

### CODE REVIEW

**Comment by mattip** [comment]
Are there tests in lib-python that currently fail that now pass with this?

**Comment by youknowone** [comment]
Sure, I started this patch from there. I will check it and leave the specific test names

**Comment by youknowone** [comment]
test command:
```
./pypy3.12-695-c -m test test_typing test_type_params test_type_aliases test_grammar test_ast test_inspect -v > result.pep695
```

edited diff to remove meaningless diff
```diff
diff -u result.orig result.pep695
--- result.orig	2026-08-19 00:36:01
+++ result.pep695	2026-08-19 01:07:35
@@ -1,14 +1,14 @@
-== PyPy 3.12.13 (83ddc70b7f97, Aug 18 2026, 15:13:54) [PyPy 8.0.0-alpha0 with GCC Apple LLVM 21.0.0 (clang-2100.1.1.101)]
+== PyPy 3.12.13 (208e492862e8, Aug 18 2026, 15:40:37) [PyPy 8.0.0-alpha0 with GCC Apple LLVM 21.0.0 (clang-2100.1.1.101)]
 == macOS-26.5-arm64-arm-64bit little-endian
 == Python build: release with_assert

@@ -685,7 +685,7 @@
 test_shadowing_nonlocal (test.test_type_params.TypeParamsAccessTest.test_shadowing_nonlocal) ... ok
 test_super (test.test_type_params.TypeParamsAccessTest.test_super) ... ok
 test_type_alias_containing_lambda (test.test_type_params.TypeParamsAccessTest.test_type_alias_containing_lambda) ... ok
-test_type_special_case (test.test_type_params.TypeParamsAccessTest.test_type_special_case) ... FAIL
+test_type_special_case (test.test_type_params.TypeParamsAccessTest.test_type_special_case) ... ok
 test_alias (test.test_type_params.TypeParamsClassScopeTest.test_alias) ... ok
 test_binding_uses_global (test.test_type_params.TypeParamsClassScopeTest.test_binding_uses_global) ... ok
 test_bound (test.test_type_params.TypeParamsClassScopeTest.test_bound) ... ok
@@ -693,14 +693,14 @@
 test_explicit_global_with_assignment (test.test_type_params.TypeParamsClassScopeTest.test_explicit_global_with_assignment) ... ok
 test_explicit_global_with_no_static_bound (test.test_type_params.TypeParamsClassScopeTest.test_explicit_global_with_no_static_bound) ... ok
 test_explicit_nonlocal (test.test_type_params.TypeParamsClassScopeTest.test_explicit_nonlocal) ... ok
-test_modified_later (test.test_type_params.TypeParamsClassScopeTest.test_modified_later) ... FAIL
-test_nested_free (test.test_type_params.TypeParamsClassScopeTest.test_nested_free) ... ERROR
+test_modified_later (test.test_type_params.TypeParamsClassScopeTest.test_modified_later) ... ok
+test_nested_free (test.test_type_params.TypeParamsClassScopeTest.test_nested_free) ... ok
 test_no_binding_uses_nonlocal (test.test_type_params.TypeParamsClassScopeTest.test_no_binding_uses_nonlocal) ... ok
 test_complex_base (test.test_type_params.TypeParamsComplexCallsTest.test_complex_base) ... ok
 test_defaults (test.test_type_params.TypeParamsComplexCallsTest.test_defaults) ... ok
 test_starargs_base (test.test_type_params.TypeParamsComplexCallsTest.test_starargs_base) ... ok
 test_disallowed_expressions (test.test_type_params.TypeParamsInvalidTest.test_disallowed_expressions) ... ok
-test_incorrect_mro_explicit_object (test.test_type_params.TypeParamsInvalidTest.test_incorrect_mro_explicit_object) ... FAIL
+test_incorrect_mro_explicit_object (test.test_type_params.TypeParamsInvalidTest.test_incorrect_mro_explicit_object) ... ok
 test_name_collisions (test.test_type_params.TypeParamsInvalidTest.test_name_collisions) ... ok
 test_name_non_collision_02 (test.test_type_params.TypeParamsInvalidTest.test_name_non_collision_02) ... ok
 test_name_non_collision_03 (test.test_type_params.TypeParamsInvalidTest.test_name_non_collision_03) ... ok
@@ -721,7 +721,7 @@
 test_no_leaky_mangling_in_module (test.test_type_params.TypeParamsManglingTest.test_no_leaky_mangling_in_module) ... ok
 test_no_mangling_in_bases (test.test_type_params.TypeParamsManglingTest.test_no_mangling_in_bases) ... ok
 test_no_mangling_in_nested_scopes (test.test_type_params.TypeParamsManglingTest.test_no_mangling_in_nested_scopes) ... ok
-test_type_params_are_mangled (test.test_type_params.TypeParamsManglingTest.test_type_params_are_mangled) ... ERROR
+test_type_params_are_mangled (test.test_type_params.TypeParamsManglingTest.test_type_params_are_mangled) ... ok
 test_nonlocal_allowed (test.test_type_params.TypeParamsNonlocalTest.test_nonlocal_allowed) ... ok
 test_nonlocal_disallowed_01 (test.test_type_params.TypeParamsNonlocalTest.test_nonlocal_disallowed_01) ... ok
 test_nonlocal_disallowed_02 (test.test_type_params.TypeParamsNonlocalTest.test_nonlocal_disallowed_02) ... ok
@@ -748,84 +748,24 @@
 test_typevartuple_02 (test.test_type_params.TypeParamsTypeVarTupleTest.test_typevartuple_02) ... ok
 test_weakrefs (test.test_type_params.TypeParamsWeakRefTest.test_weakrefs) ... ok
 
-======================================================================
-ERROR: test_nested_free (test.test_type_params.TypeParamsClassScopeTest.test_nested_free)
 ----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 703, in test_nested_free
-    ns = run_code("""
-         ^^^^^^^^^^^^
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/support/__init__.py", line 665, in run_code
-    exec(textwrap.dedent(code), ns)
-SystemError: unexpected internal exception (please report a bug): <KeyError object at 0x1058060d8>; internal traceback was dumped to stderr
-
-======================================================================
-ERROR: test_type_params_are_mangled (test.test_type_params.TypeParamsManglingTest.test_type_params_are_mangled)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 840, in test_type_params_are_mangled
-    ns = run_code("""
-         ^^^^^^^^^^^^
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/support/__init__.py", line 665, in run_code
-    exec(textwrap.dedent(code), ns)
-  File "<string>", line 4, in <module>
-  File "<string>", line 4, in Foo
-  File "<string>", line 5, in Foo
-NameError: name '_Foo__T' is not defined
-
-======================================================================
-FAIL: test_type_special_case (test.test_type_params.TypeParamsAccessTest.test_type_special_case)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 504, in test_type_special_case
-    self.assertEqual(type.__type_params__, ())
-         ^^^^^^^^^^^
-AssertionError: <getset_descriptor object at 0x0000000105ae7c90> != ()
+Ran 90 tests in 0.018s
 
-======================================================================
-FAIL: test_modified_later (test.test_type_params.TypeParamsClassScopeTest.test_modified_later)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 607, in test_modified_later
-    self.assertIs(X.foo.__type_params__[0].__bound__, float)
-         ^^^^^^^^
-AssertionError: <class 'int'> is not <class 'float'>
-
-======================================================================
-FAIL: test_incorrect_mro_explicit_object (test.test_type_params.TypeParamsInvalidTest.test_incorrect_mro_explicit_object)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 153, in test_incorrect_mro_explicit_object
-    class My[X](object): ...
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 153, in My
-    class My[X](object): ...
-TypeError: cycle among base classes: object < Generic < object
-
-During handling of the above exception, another exception occurred:
-
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_params.py", line 152, in test_incorrect_mro_explicit_object
-    with self.assertRaisesRegex(TypeError, r"\(MRO\) for bases object, Generic"):
-AssertionError: "\(MRO\) for bases object, Generic" does not match "cycle among base classes: object < Generic < object"
-
-----------------------------------------------------------------------
-Ran 90 tests in 0.056s
-
-FAILED (failures=3, errors=2)
-0:00:00 load avg: 106.64 [2/6/1] test_type_params failed (2 errors, 3 failures)
-0:00:00 load avg: 106.64 [3/6/1] test_type_aliases
+OK
+0:00:00 load avg: 21.29 [2/6] test_type_params passed
+0:00:00 load avg: 21.29 [3/6] test_type_aliases
 test_attributes_with_exec (test.test_type_aliases.TypeAliasConstructorTest.test_attributes_with_exec) ... ok
 test_basic (test.test_type_aliases.TypeAliasConstructorTest.test_basic) ... ok
 test_errors (test.test_type_aliases.TypeAliasConstructorTest.test_errors) ... ok
-test_generic (test.test_type_aliases.TypeAliasConstructorTest.test_generic) ... FAIL
+test_generic (test.test_type_aliases.TypeAliasConstructorTest.test_generic) ... ok
 test_keywords (test.test_type_aliases.TypeAliasConstructorTest.test_keywords) ... ok
 test_not_generic (test.test_type_aliases.TypeAliasConstructorTest.test_not_generic) ... ok
 test_pickling (test.test_type_aliases.TypeAliasPickleTest.test_pickling) ... ok
 test_pickling_local (test.test_type_aliases.TypeAliasPickleTest.test_pickling_local) ... ok
-test_immutable (test.test_type_aliases.TypeAliasTypeTest.test_immutable) ... FAIL
+test_immutable (test.test_type_aliases.TypeAliasTypeTest.test_immutable) ... ok
 test_module (test.test_type_aliases.TypeAliasTypeTest.test_module) ... ok
-test_no_subclassing (test.test_type_aliases.TypeAliasTypeTest.test_no_subclassing) ... FAIL
-test_union (test.test_type_aliases.TypeAliasTypeTest.test_union) ... FAIL
+test_no_subclassing (test.test_type_aliases.TypeAliasTypeTest.test_no_subclassing) ... ok
+test_union (test.test_type_aliases.TypeAliasTypeTest.test_union) ... ok
 test_alias_access_01 (test.test_type_aliases.TypeParamsAccessTest.test_alias_access_01) ... ok
 test_alias_access_02 (test.test_type_aliases.TypeParamsAccessTest.test_alias_access_02) ... ok
 test_alias_access_03 (test.test_type_aliases.TypeParamsAccessTest.test_alias_access_03) ... ok
@@ -833,99 +773,21 @@
 test_alias_value_02 (test.test_type_aliases.TypeParamsAliasValueTest.test_alias_value_02) ... ok
 test_alias_value_03 (test.test_type_aliases.TypeParamsAliasValueTest.test_alias_value_03) ... ok
 test_alias_value_04 (test.test_type_aliases.TypeParamsAliasValueTest.test_alias_value_04) ... ok
-test_raising (test.test_type_aliases.TypeParamsAliasValueTest.test_raising) ... FAIL
+test_raising (test.test_type_aliases.TypeParamsAliasValueTest.test_raising) ... ok
 test_recursive_repr (test.test_type_aliases.TypeParamsAliasValueTest.test_recursive_repr) ... ok
-test_repr (test.test_type_aliases.TypeParamsAliasValueTest.test_repr) ... FAIL
-test_subscripting (test.test_type_aliases.TypeParamsAliasValueTest.test_subscripting) ... FAIL
+test_repr (test.test_type_aliases.TypeParamsAliasValueTest.test_repr) ... ok
+test_subscripting (test.test_type_aliases.TypeParamsAliasValueTest.test_subscripting) ... ok
 test_exec_with_unusual_globals (test.test_type_aliases.TypeParamsExoticGlobalsTest.test_exec_with_unusual_globals) ... ok
 test_name_collisions (test.test_type_aliases.TypeParamsInvalidTest.test_name_collisions) ... ok
 test_name_non_collision_02 (test.test_type_aliases.TypeParamsInvalidTest.test_name_non_collision_02) ... ok
 test_name_non_collision_03 (test.test_type_aliases.TypeParamsInvalidTest.test_name_non_collision_03) ... ok
 
-======================================================================
-FAIL: test_generic (test.test_type_aliases.TypeAliasConstructorTest.test_generic)
 ----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 215, in test_generic
-    self.assertIs(type(TA[int]), types.GenericAlias)
-         ^^^^^^^^
-AssertionError: <class 'typing._GenericAlias'> is not <class '_pypy_generic_alias.GenericAlias'>
+Ran 27 tests in 0.010s
 
-======================================================================
-FAIL: test_immutable (test.test_type_aliases.TypeAliasTypeTest.test_immutable)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 249, in test_immutable
-    with self.assertRaises(TypeError):
-AssertionError: TypeError not raised
-
-======================================================================
-FAIL: test_no_subclassing (test.test_type_aliases.TypeAliasTypeTest.test_no_subclassing)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 253, in test_no_subclassing
-    with self.assertRaisesRegex(TypeError, "not an acceptable base type"):
-AssertionError: TypeError not raised
-
-======================================================================
-FAIL: test_union (test.test_type_aliases.TypeAliasTypeTest.test_union)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 261, in test_union
-    self.assertIsInstance(union, types.UnionType)
-         ^^^^^^^^^^^^^^^^
-AssertionError: typing.Union[Alias1, Alias2] is not an instance of <class '_pypy_generic_alias.UnionType'>
-
-======================================================================
-FAIL: test_raising (test.test_type_aliases.TypeParamsAliasValueTest.test_raising)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 178, in test_raising
-    MissingName.__value__
-                ^^^^^^^^^
-  File "/Users/youknowone/Projects/pypy-2/lib_pypy/_pypy_typing.py", line 142, in __get__
-    value = getattr(instance, f"__evaluate_{self._name}__")()
-            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 173, in MissingName
-    type MissingName = list[_My_X]
-                            ^^^^^
-NameError: free variable '_My_X' referenced before assignment in enclosing scope
-
-During handling of the above exception, another exception occurred:
-
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 174, in test_raising
-    with self.assertRaisesRegex(
-AssertionError: "cannot access free variable '_My_X' where it is not associated with a value" does not match "free variable '_My_X' referenced before assignment in enclosing scope"
-
-======================================================================
-FAIL: test_repr (test.test_type_aliases.TypeParamsAliasValueTest.test_repr)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 150, in test_repr
-    self.assertEqual(repr(VeryGeneric[int, bytes, str, [float, object]]),
-         ^^^^^^^^^^^
-AssertionError: "VeryGeneric[int, bytes, str, [<class 'float'>, <class 'object'>]]" != 'VeryGeneric[int, bytes, str, [float, object]]'
-- VeryGeneric[int, bytes, str, [<class 'float'>, <class 'object'>]]
-?                               --------     --  --------      --
-+ VeryGeneric[int, bytes, str, [float, object]]
-
-
-======================================================================
-FAIL: test_subscripting (test.test_type_aliases.TypeParamsAliasValueTest.test_subscripting)
-----------------------------------------------------------------------
-Traceback (most recent call last):
-  File "/Users/youknowone/Projects/pypy-2/lib-python/3/test/test_type_aliases.py", line 135, in test_subscripting
-    self.assertIsInstance(specialized, types.GenericAlias)
-         ^^^^^^^^^^^^^^^^
-AssertionError: Generic[int] is not an instance of <class '_pypy_generic_alias.GenericAlias'>
-
-----------------------------------------------------------------------
-Ran 27 tests in 0.014s
-
-FAILED (failures=7)
-0:00:00 load avg: 106.64 [3/6/2] test_type_aliases failed (7 failures)
-0:00:00 load avg: 106.64 [4/6/2] test_grammar
+OK
+0:00:00 load avg: 21.29 [3/6] test_type_aliases passed
+0:00:00 load avg: 21.29 [4/6] test_grammar
 test_additive_ops (test.test_grammar.GrammarTests.test_additive_ops) ... ok
 test_annotations_inheritance (test.test_grammar.GrammarTests.test_annotations_inheritance) ... ok
 test_assert (test.test_grammar.GrammarTests.test_assert) ... ok
@@ -1063,11 +925,11 @@
 AssertionError: Regex didn't match: '"is" with \'int\' literal' not found in '"is" with a literal. Did you mean "=="?'
 
 ----------------------------------------------------------------------
-Ran 79 tests in 0.087s
+Ran 79 tests in 0.026s
 
 FAILED (failures=1, errors=3)
-0:00:01 load avg: 106.64 [4/6/3] test_grammar failed (3 errors, 1 failure)
-0:00:01 load avg: 106.64 [5/6/3] test_ast
+0:00:00 load avg: 21.29 [4/6/1] test_grammar failed (3 errors, 1 failure)
+0:00:00 load avg: 21.29 [5/6/1] test_ast
 test_bad_integer (test.test_ast.test_ast.ASTHelpers_Test.test_bad_integer) ... ok
 test_copy_location (test.test_ast.test_ast.ASTHelpers_Test.test_copy_location) ... ok
 test_dump (test.test_ast.test_ast.ASTHelpers_Test.test_dump) ... ok
@@ -1235,11 +1097,11 @@
 test_old_constant_nodes (test.test_ast.test_ast.NodeVisitorTests.test_old_constant_nodes) ... ok
 
 ----------------------------------------------------------------------
-Ran 165 tests in 0.432s
+Ran 165 tests in 0.216s
 
 OK (skipped=8)
-0:00:01 load avg: 106.64 [5/6/3] test_ast passed
-0:00:01 load avg: 106.64 [6/6/3] test_inspect.test_inspect
+0:00:01 load avg: 21.29 [5/6/1] test_ast passed
+0:00:01 load avg: 21.29 [6/6/1] test_inspect.test_inspect
 test__all__ (test.test_inspect.test_inspect.IsTestBase.test__all__) ... ok
 test_class_async_method (test.test_inspect.test_inspect.TestBlockComments.test_class_async_method) ... ok
 test_class_method (test.test_inspect.test_inspect.TestBlockComments.test_class_method) ... ok
@@ -1740,20 +1602,19 @@
 AssertionError: None != 'measured in kilowatts'
 
 ----------------------------------------------------------------------
-Ran 306 tests in 0.709s
+Ran 306 tests in 0.564s
 
 FAILED (failures=8, errors=5, skipped=11)
-0:00:02 load avg: 106.64 [6/6/4] test_inspect.test_inspect failed (5 errors, 8 failures)
+0:00:01 load avg: 21.29 [6/6/2] test_inspect.test_inspect failed (5 errors, 8 failures)
 
 == Tests result: FAILURE ==
 
-4 tests failed:
-    test_grammar test_inspect.test_inspect test_type_aliases
-    test_type_params
+2 tests failed:
+    test_grammar test_inspect.test_inspect
 
-2 tests OK.
+4 tests OK.
 
 Result: FAILURE
 ```

I checked tests are fixed, but not yet fully reviewed the code diff

**Comment by youknowone** [comment]
@mattip converted to ready

File: pypy/interpreter/pyopcode.py [hunk 1]
```diff
@@ -677,9 +677,12 @@ def raise_exc_unbound(self, varindex):
                         "local variable '%s' referenced before assignment",
                         varname)
         else:
+            # Match CPython 3.12+ wording (see ceval.c:_PyEval_EvalFrameDefault
+            # unbound free variable message).
```
Comment (youknowone) [inline pypy/interpreter/pyopcode.py]:
```suggestion
```

File: lib_pypy/_pypy_typing.py [hunk 2]
```diff
@@ -366,7 +394,27 @@ def __init_subclass__(cls, **kwargs):
         raise TypeError("type 'typing.TypeVarTuple' is not an acceptable base type")
 
 
-class TypeAliasType(_Immutable, _PickleUsingNameMixin):
+class _TypeAliasTypeMeta(type):
```
Comment (youknowone) [inline lib_pypy/_pypy_typing.py]: probably this must not be only _TypeAliasTypeMeta. is there any metaclass for this usage?

File: lib_pypy/_pypy_typing.py [hunk 3]
```diff
@@ -379,6 +427,22 @@ class TypeAliasType(_Immutable, _PickleUsingNameMixin):
         # and __value__ = tuple[float, float]
     """
 
+    __module__ = 'typing'
+
+    # _PickleUsingNameMixin has no __slots__, so inheriting from it would add
```
Comment (youknowone) [inline lib_pypy/_pypy_typing.py]: this is looking suspicious. is this good decision?

File: lib_pypy/_pypy_typing.py [hunk 4]
```diff
@@ -387,56 +451,96 @@ def __init__(self, name, value, *, type_params=()):
...
-        return self._name
+        return self.__name_value
 
     @property
     def __type_params__(self):
-        return self._type_params
+        return self.__type_params_value
+
+    def __getattribute__(self, name):
+        # The class itself must expose TypeAliasType.__module__ == 'typing',
```
Comment (youknowone) [inline lib_pypy/_pypy_typing.py]: need to check this is right way

File: pypy/objspace/std/typeobject.py [hunk 5]
```diff
@@ -990,10 +1002,25 @@ def _store_type_in_classcell(space, w_type, w_classcell, dict_w):
...
+                    space.type(w_classcell))
     del dict_w['__classcell__']
 
+def _validate_classdictcell(space, w_classdictcell, dict_w):
+    from pypy.interpreter.nestedscope import Cell
+    if not isinstance(w_classdictcell, Cell):
+        raise oefmt(space.w_TypeError,
+                    "__classdictcell__ must be a nonlocal cell, not %R",
+                    space.type(w_classdictcell))
+    del dict_w['__classdictcell__']
```
Comment (cfbolz) [inline pypy/objspace/std/typeobject.py]: this is a strange helper function, it's called "validate", but it also mutates the dict_w. I think the del should move to the caller. if you want the rpython type knowledge to stick, you could return w_classdictcell on success

File: pypy/objspace/std/typeobject.py [hunk 6]
```diff
@@ -990,10 +1002,25 @@ def _store_type_in_classcell(space, w_type, w_classcell, dict_w):
...
+
+def _update_classdictcell(space, w_classdictcell, w_type):
+    from pypy.interpreter.nestedscope import Cell
+    # Keep this check at the call boundary: the RPython annotator does not
+    # propagate the type narrowing performed by _validate_classdictcell().
+    if isinstance(w_classdictcell, Cell):
```
Comment (cfbolz) [inline pypy/objspace/std/typeobject.py]: this should not be an if, but an assert

File: pypy/objspace/std/typeobject.py [hunk 7]
```diff
@@ -1756,22 +1794,23 @@ def mro_blockinglist(candidate, orderlists):
     return None # good candidate
 
 def mro_error(space, orderlists):
-    cycle = []
     candidate = orderlists[-1][0]
     if candidate in orderlists[-1][1:]:
         # explicit error message for this specific case
         raise oefmt(space.w_TypeError, "duplicate base class %N", candidate)
-    while candidate not in cycle:
```
Comment (cfbolz) [inline pypy/objspace/std/typeobject.py:1764]: we need to be less close to CPython here. PyPy has strictly more information in its error messages for these mro problems. I am fine with adapting more to the CPython wording, but I definitely want to keep the more precise cycle computation that PyPy has.

**Comment by youknowone** [comment]
@cfbolz Thank you so much for the review!

File: pypy/objspace/std/typeobject.py (hunk 5 above)
Comment (youknowone) [reply pypy/objspace/std/typeobject.py]: Thanks! This part becomes more simple

File: pypy/objspace/std/typeobject.py (hunk 7 above)
Comment (youknowone) [reply pypy/objspace/std/typeobject.py:1764]: Thanks! I had to look this more careful. I also applied similar policy to: https://github.com/pypy/pypy/pull/5561/changes#diff-653d3f7ea72dd58c37350bd48be6c833fa37402d389b4692b568052e3a6111c9R89-R90

File: lib_pypy/_pypy_typing.py (hunk 2 above)
Comment (youknowone) [reply lib_pypy/_pypy_typing.py]: this is now `_ImmutableTypeMeta`

**Comment by mattip** [comment]
I tried this out and it seems to work well. Merging. Thanks @youknowone
